src/Evolution/Systems/GrMhd/ValenciaDivClean/BoundaryCorrections/Hll.cpp

```diff
@@ -61,26 +68,116 @@ double Hll::dg_package_data(
+          equation_of_state
+              .specific_internal_energy_from_density_and_temperature(
+                  rest_mass_density, temperature, electron_fraction);
+      Scalar<DataVector> pressure =
+          equation_of_state.pressure_from_density_and_energy(
+              rest_mass_density, specific_internal_energy, electron_fraction);
+      Scalar<DataVector> specific_enthalpy =
+          hydro::relativistic_specific_enthalpy(
+              rest_mass_density, specific_internal_energy, pressure);
+      Scalar<DataVector> sound_speed_squared =
+          equation_of_state.sound_speed_squared_from_density_and_temperature(
+              rest_mass_density, temperature, electron_fraction);
+
+      // Get v0 from Font
+      Scalar<DataVector> v0 = tilde_d;
+      dot_product(make_not_null(&v0), spatial_velocity, normal_covector);
+      get(v0) *= get(lapse);
+
+      // Get v^2
+      Scalar<DataVector> v_squared = tilde_d;
+      dot_product(make_not_null(&v_squared), spatial_velocity,
+                  spatial_velocity_one_form);
+
+      for (size_t i = 0; i < get(rest_mass_density).size(); i++) {
+        if (get(sound_speed_squared)[i] > 1.0) {
+          get(sound_speed_squared)[i] = 1.0;
+        }
+        if (get(v_squared)[i] > 1.0 - 1.e-8) {
+          get(v_squared)[i] = 1.0 - 1.e-8;
+        }
+        if (get(sound_speed_squared)[i] < 0) {
+          get(sound_speed_squared)[i] = 0;
+        }
+        if (get(v_squared)[i] < 0) {
+          get(v_squared)[i] = 0;
+        }
+      }
+
+      // Calculate characteristic speeds in inertial frame
+      DataVector discriminant = get(sound_speed_squared);
+      discriminant = get(sound_speed_squared) * (1.0 - get(v_squared)) *
+                     (get(lapse) * get(lapse) *
+                          (1.0 - get(v_squared) * get(sound_speed_squared)) -
+                      get(v0) * get(v0) * (1.0 - get(sound_speed_squared)));
+      for (size_t i = 0; i < get(rest_mass_density).size(); i++) {
+        if (discriminant[i] < 0.0) {
+          discriminant[i] = 0.0;
+        }
+      }
+
+      discriminant = sqrt(discriminant);
       get(*packaged_largest_outgoing_char_speed) =
-          get(lapse) - get(shift_dot_normal) - get(*normal_dot_mesh_velocity);
+          (get(v0) * (1.0 - get(sound_speed_squared)) + discriminant) /
+              (1.0 - get(v_squared) * get(sound_speed_squared)) -
+          get(shift_dot_normal);
       get(*packaged_largest_ingoing_char_speed) =
-          -get(lapse) - get(shift_dot_normal) - get(*normal_dot_mesh_velocity);
-    } else {
+          (get(v0) * (1.0 - get(sound_speed_squared)) - discriminant) /
+              (1.0 - get(v_squared) * get(sound_speed_squared)) -
+          get(shift_dot_normal);
+
+      for (size_t i = 0; i < get(rest_mass_density).size(); i++) {
+        if (get(rest_mass_density)[i] < 1.e-08) {
+          double alpha = 0.0;  // get(rest_mass_density)[i]/1.e-08;
```

Comment:
confusion... why is alpha 0 and then we are doing math with alpha?

Reply:
@ffoucart do we still need this?

Reply:
It used to be necessary for stability of the atmosphere. Effectively, it makes the code use the speed of light as characteristic speed at low density. I have not tried to remove that correction though, and it's probably not a great idea to keep the hard-coded 1.e-8 in the final version.
My preference, if possible, would be to get rid of the alpha, just set the speeds to the speed of light at low density, and have the density threshold as an specifiable input that we can test on merger simulations and maybe lower / remove at a later time if it proves unnecessary.